src/artifact.rs

```diff
@@ -447,10 +447,34 @@ impl Artifact {
 
     }
 
+    /// Get set of non-import declarations that have not been defined. This must be an empty set in
+    /// order to `emit` the artifact.
+    pub fn undefined_symbols(&self) -> Vec<String> {
+        let mut syms = Vec::new();
+        for (&name, &decl) in self.declarations.iter() {
+            match decl {
+                Decl::FunctionImport => {},
+                Decl::DataImport => {},
+                Decl::Function { .. } | Decl::Data { .. } | Decl::CString { .. }=> {
+                    if self.definitions.iter().find(|def| def.name == name).is_none() {
```

Comment:
What would you think about adding a `defined` flag to the `Function`/`Data`/`CString` arms of `Decl`, and having the `define` function set it to true? That way we wouldn't have to scan the definitions for each import here.